src/core/primitives.rs

```diff
@@ -451,3 +451,52 @@ fn refresh_coordinate_keys(
         })
         .collect::<RevisionVec<_, _>>()
 }
+
+/// Attempts opening the Covercrypt encapsulation using the given USK. Returns
+/// the encapsulated key and associated rights upon success, otherwise returns `None`.
+pub fn full_decaps(
+    usk: &UserSecretKey,
+    encapsulation: &XEnc,
+) -> Result<Option<Vec<(Right, Secret<SHARED_SECRET_LENGTH>)>>, Error> {
+    // A = ⊙ _i (α_i. c_i)
+    let A = usk
+        .id
+        .iter()
+        .zip(encapsulation.c.iter())
+        .map(|(marker, trap)| trap * marker)
+        .fold(EcPoint::identity(), |mut acc, elt| {
+            acc = &acc + &elt;
+            acc
+        });
+
+    for enc in &encapsulation.encapsulations {
+        let mut rights_list: Vec<(Right, Secret<SHARED_SECRET_LENGTH>)> = Vec::new();
+        for secret in usk.secrets.flat_iter() {
```

Comment:
This is misleading, I had to read `l.496` to understand : de-structure the secret instead
```rust
for (right, secret) in usk.secrets.flat_iter() {
```